graph/api/api.go

```diff
@@ -35,20 +37,20 @@ func GraphNamespaces(business *business.Layer, o graph.Options) (code int, confi
 }
 
 // graphNamespacesIstio provides a test hook that accepts mock clients
-func graphNamespacesIstio(business *business.Layer, prom *prometheus.Client, o graph.Options) (code int, config interface{}) {
+func graphNamespacesIstio(ctx context.Context, business *business.Layer, prom *prometheus.Client, o graph.Options) (code int, config interface{}) {
 
 	// Create a 'global' object to store the business. Global only to the request.
 	globalInfo := graph.NewAppenderGlobalInfo()
 	globalInfo.Business = business
 
 	trafficMap := istio.BuildNamespacesTrafficMap(o.TelemetryOptions, prom, globalInfo)
-	code, config = generateGraph(trafficMap, o)
+	code, config = generateGraph(ctx, trafficMap, o, business)
```

Comment:
OK, I'll see if I can move all this to an appender or finalizer.

Reply:
Feel free to ping me if you want to talk about this.   We have 3 "finalizer" appenders at the moment.  Two of them always run (not optional) and this could be like that as well.  The third is the one just added recently, called "labeler", and it runs optionally.  So if you wanted to keep the health stuff optional you could do that easily, it's just another appender, not a new option.

Reply:
I was looking at it and this seems to need the very final graph.

It turns out that boxes can have health. Specifically, the "app boxes", like the "reviews" in the bookinfo sample app:

![image](https://user-images.githubusercontent.com/23639005/158090743-ac08e231-960a-4069-a4e6-d0037953b4bf.png)

Healt is calculated differently: for the inner nodes, workload health is fetched (despite they are "app" nodes) and for the enclosing app box, app health is fetched. Regardless about how health would be calculated, I do need a way to carry metadata about app boxes in the traffic map to be able to attach health data to them. That probably would imply moving around or replicating the logic for resolving app compound boxes from the `generateGraph` function to the `BuildTrafficMap` functions :confused:

Reply:
I understand what you are saying.  Let me take a look at the code and see if there is something that makes sense.  Perhaps the app box health can be inherited from the children, like some other app box information.   I really would prefer to keep the config generation from having to perform additional graph generation tasks.

Reply:
Thinking a bit, there may be ways this can be done in a standard health finalizer appender.  What do you think about this approach:  When you encounter an app node in the TrafficMap, query for the app health.  Store it on the app node in the metadata.  For an "App graph", this would be the same thing we do already.  But for a "Versioned App" graph, we'd store the workload health AND the app health.  We'd cache the app health in the appender code, so we could assign it to every versioned-app node for that app.  Then, for an app box node, the config generator could just inherit the app health from any one of the child app nodes.

Reply:
OK, I think that's doable.

With that in mind, it is probably better to keep the `includeHealth` flag (which would become an optional appender). Otherwise, it is going to be a lot of work to fix all tests in `api_test.go`. Making health mandatory would mean that mocks are needed on almost all tests on that file :-/